staging/src/k8s.io/component-base/featuregate/testing/feature_gate.go

```diff
@@ -18,18 +18,54 @@ package testing
 
 import (
 	"fmt"
-	"testing"
+	"strings"
+	"sync"
 
 	"k8s.io/component-base/featuregate"
 )
 
-// SetFeatureGateDuringTest sets the specified gate to the specified value, and returns a function that restores the original value.
-// Failures to set or restore cause the test to fail.
+var (
+	overwrittenFeatureFlags map[featuregate.Feature]string
+	lock                    sync.Mutex
+)
+
+func init() {
+	overwrittenFeatureFlags = map[featuregate.Feature]string{}
+}
+
+// SetFeatureGateDuringTest sets the specified gate to the specified value for duration of the test.
+// Fails when it detects second call to the same flag or is unable to set or restore feature flag.
+// Returns empty cleanup function to maintain the old function signature that uses defer.
+// TODO: Remove defer from calls to SetFeatureGateDuringTest and update hack/verify-test-featuregates.sh when we can do large scale code change.
+//
+// WARNING: Can leak set variable when called in test calling t.Parallel(), however second attempt to set the same feature flag will cause fatal.
 //
 // Example use:
 //
 // defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.<FeatureName>, true)()
-func SetFeatureGateDuringTest(tb testing.TB, gate featuregate.FeatureGate, f featuregate.Feature, value bool) func() {
+func SetFeatureGateDuringTest(tb testingTB, gate featuregate.FeatureGate, f featuregate.Feature, value bool) func() {
+	tb.Helper()
+	lock.Lock()
+	defer lock.Unlock()
+	cleanup := setFeatureGateDuringTest(tb, gate, f, value)
+
+	tb.Cleanup(func() {
+		tb.Helper()
+		lock.Lock()
+		defer lock.Unlock()
+		cleanup()
+	})
+	return func() {
+	}
+}
+
+func setFeatureGateDuringTest(tb testingTB, gate featuregate.FeatureGate, f featuregate.Feature, value bool) func() {
+	tb.Helper()
+	previousTestName, alreadyOverwritten := overwrittenFeatureFlags[f]
+	if alreadyOverwritten && !strings.HasPrefix(tb.Name(), previousTestName) {
```

Comment:
I think a prefix check will treat parallel "TestFoo" and "TestFooBar" as ok, when really, we only want to allow subtests of `TestFoo` like `TestFoo/Bar`, right?

Should this check exactly equality or prefix + "/"?

```suggestion
	if alreadyOverwritten && tb.Name() != previousTestName && !strings.HasPrefix(tb.Name(), previousTestName + "/") {
```

Reply:
Right, sad that I missed that :(

Reply:
Fixed and added a test. Still user can use "/" in the test name, but that should be much less frequent.